### PR TITLE
Feature ghn api

### PR DESCRIPTION
**Pull request by tathai2610**
# Description 
- Add GhnClient class with the responsibility of communicating between current website and GHN 
- Opening a shop now will automatically register a new shop on GHN server
- Creating orders now will automatically create orders on GHN server
- Update database diagram 
  - Add `shipping_code` in City, District, Ward
  - Add `code` to Shop and Order 
 
# Screenshot
### Database diagram (updated)
db_diagram

### CODE REVIEW

File: app/apis/ghn_client.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
+require 'json'
+
+class GhnClient 
+  include HTTParty
+  base_uri 'dev-online-gateway.ghn.vn/shiip/public-api'
+  headers token: ENV["GHN_API_KEY"], "Content-Type": "application/json"
+  debug_output $stdout
```
Comment (nathan-phan-goldenowl) [inline app/apis/ghn_client.rb:7]: remove this after debugging

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,3 +1,5 @@
+require Rails.root.join('app', 'apis', 'ghn_client.rb')
```
Comment (nathan-phan-goldenowl) [inline app/controllers/application_controller.rb:1]: why you need to add this line here?

File: app/controllers/shops_controller.rb [hunk 3]
```diff
@@ -4,23 +4,27 @@ class ShopsController < ApplicationController
   before_action :check_shop_exist, only: [:new, :create]
 
   def show 
-    @products = @shop.products 
+    @products = Product.available.where(shop: @shop)
+    @products = @shop.products if @user == current_user
```
Comment (nathan-phan-goldenowl) [inline app/controllers/shops_controller.rb:8]: This makes almost 2 duplicate queries.
Use pundit scope here to filter out product scope to show

File: app/jobs/update_orders_status_job.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
+class UpdateOrdersStatusJob < ApplicationJob
+  queue_as :default
+
+  def perform(order)
+    # get order info from ghn api
+    response = GhnClient.new.get_order_info(order.code)
+    order_repsonse_status = response["data"]["status"]
+    byebug
+    # puts "response: " + response["data"]["status"]
```
Comment (nathan-phan-goldenowl) [inline app/jobs/update_orders_status_job.rb:9]: remove these lines

File: app/jobs/update_orders_status_job.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+    Order::STATUSES.each do |k, v|
+      if v.include?(order_repsonse_status.to_sym)
+        # update current order status unless new status is the same as current status
+        unless k.to_s == order.status_name
+          # puts "before: " + order.status
+          order.status = k.to_s
+          order.save
+          # puts "after: " + order.status
```
Comment (nathan-phan-goldenowl) [inline app/jobs/update_orders_status_job.rb:18]: remove debugging comments

File: app/apis/ghn_client.rb (hunk 1 above)
Comment (tathai2610) [reply app/apis/ghn_client.rb:7]: noted and updated

File: app/jobs/update_orders_status_job.rb (hunk 4 above)
Comment (tathai2610) [reply app/jobs/update_orders_status_job.rb:9]: noted and updated

File: app/jobs/update_orders_status_job.rb (hunk 5 above)
Comment (tathai2610) [reply app/jobs/update_orders_status_job.rb:18]: noted and updated

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (tathai2610) [reply app/controllers/application_controller.rb:1]: on the second look, I now realize its unnecessary, it is then removed

File: app/controllers/shops_controller.rb (hunk 3 above)
Comment (tathai2610) [reply app/controllers/shops_controller.rb:8]: noted and updated

**Pull request merged**